lib/plugins/aws/deploy/lib/ensure-valid-bucket-exists.js

```diff
@@ -0,0 +1,147 @@
+'use strict';
+
+const ServerlessError = require('../../../../serverless-error');
+const _ = require('lodash');
+const { log, legacy } = require('@serverless/utils/log');
+
+module.exports = {
+  async ensureValidBucketExists() {
+    legacy.log('Ensuring that deployment bucket exists');
+
+    // Validate that custom deployment bucket exists and has proper location
+    if (this.serverless.service.provider.deploymentBucket) {
+      let result;
+      try {
+        result = await this.provider.request('S3', 'getBucketLocation', {
+          Bucket: this.bucketName,
+        });
+      } catch (err) {
+        throw new ServerlessError(
+          `Could not locate deployment bucket. Error: ${err.message}`,
+          'DEPLOYMENT_BUCKET_NOT_FOUND'
+        );
+      }
+
+      if (result.LocationConstraint === '') result.LocationConstraint = 'us-east-1';
+      if (result.LocationConstraint === 'EU') result.LocationConstraint = 'eu-west-1';
+      if (result.LocationConstraint !== this.provider.getRegion()) {
+        throw new ServerlessError(
+          'Deployment bucket is not in the same region as the lambda function',
+          'DEPLOYMENT_BUCKET_INVALID_REGION'
+        );
+      }
+      // If above is satisfied, then custom S3 bucket is valid
+      return;
+    }
+
+    // Ensure to set bucket name if it can be resolved
+    // Result of this operation will determine how further validation will be performed
+    try {
+      await this.setBucketName();
+    } catch (err) {
+      // If there is a validation error with expected message, it means that logical resource for
+      // S3 bucket does not exist and we want to proceed with handling that situation
+      if (
+        err.providerError.code !== 'ValidationError' ||
+        !err.message.includes('does not exist for stack')
+      ) {
+        throw err;
+      }
+    }
+
+    // If bucket name is set, it means it's defined as a part of CloudFormation template (custom bucket case was handled by logic above)
+    if (this.bucketName) {
+      if (!(await this.checkIfBucketExists(this.bucketName))) {
+        // It means that bucket was removed manually but is still a part of the CloudFormation stack, we cannot manually fix it
```

Comment:
I double checked if we can force CloudFormation deploy with the original template, but as expected, the update is rejected because "nothing changed"